app/repositories/ingame_logins.py

```diff
@@ -0,0 +1,213 @@
+) -> Optional[dict[str, Any]]:
+    """Fetch a login entry from the database."""
+    if (
+        id is None
+        and userid is None
+        and ip is None
+        and osu_ver is None
+        and osu_stream is None
+    ):
+        raise ValueError("Must provide at least one parameter.")
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE id = COALESCE(:id, id)
+           AND userid = COALESCE(:userid, userid)
+           AND ip = COALESCE(:ip, ip)
+           AND osu_ver = COALESCE(:osu_ver, osu_ver)
+           AND osu_stream = COALESCE(:osu_stream, osu_stream)
+    """
+    params = {
+        "id": id,
+        "userid": userid,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    return dict(rec) if rec is not None else None
+
+
+async def fetch_count(
+    userid: Optional[int] = None,
+    ip: Optional[str] = None,
+    osu_ver: Optional[str] = None,
+    osu_stream: Optional[str] = None,
+) -> int:
+    """Fetch the number of logins in the database."""
+    query = """\
+        SELECT COUNT(*) AS count
+          FROM ingame_logins
+        WHERE userid = COALESCE(:userid, userid)
+          AND ip = COALESCE(:ip, ip)
+          AND osu_ver = COALESCE(:osu_ver, osu_ver)
+          AND osu_stream = COALESCE(:osu_stream, osu_stream)
+
+    """
+    params = {
+        "userid": userid,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return rec["count"]
+
+
+async def fetch_many(
+    userid: Optional[int] = None,
+    ip: Optional[str] = None,
+    osu_ver: Optional[str] = None,
+    osu_stream: Optional[str] = None,
+    page: Optional[int] = None,
+    page_size: Optional[int] = None,
+) -> list[dict[str, Any]]:
+    """Fetch a list of logins from the database."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE userid = COALESCE(:userid, userid)
+           AND ip = COALESCE(:ip, ip)
+           AND osu_ver = COALESCE(:osu_ver, osu_ver)
+           AND osu_stream = COALESCE(:osu_stream, osu_stream)
+    """
+    params = {
+        "userid": userid,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+
+    if page is not None and page_size is not None:
+        query += """\
+            LIMIT :limit
+           OFFSET :offset
+        """
+        params["limit"] = page_size
+        params["offset"] = (page - 1) * page_size
+
+    recs = await app.state.services.database.fetch_all(query, params)
+    return [dict(rec) for rec in recs]
+
+
+async def update(
+    id: int,
+    userid: Optional[int] = None,
+    ip: Optional[str] = None,
+    osu_ver: Optional[str] = None,
+    osu_stream: Optional[str] = None,
```

Comment:
do we have an `Unset` pattern in bancho.py repositories? I can't remember if I ever implemented that here

if not we might want to figure it out sooner than later so that we can differentiate between nulls and unset values

Reply:
we don't